File: redqueen-backend/RedQueen/Services/MqttService.cs
using System;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using MQTTnet;
using MQTTnet.Client;
using MQTTnet.Extensions.ManagedClient;
using MQTTnet.Formatter;
using MQTTnet.Protocol;
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;
using RedQueen.Data.Models.Db;
using RedQueen.Data.Models.Dto;

namespace RedQueen.Services
{
    public interface IMqttService : IDisposable
    {
        event MqttMessageReceivedEventHandler MessageReceived;
        string Host { get; }
        bool IsDisposed { get; }
        bool AutoDiscoverEnabled { get; }
        ManagedMqttClientOptions GetOptions();
        Task StartPublisher();
        Task StopPublisher();
        Task StartSubscriber();
        Task StopSubscriber();
        Task SubscribeTopic(MqttTopic topic);
        Task SubscribeAllTopics();
        Task StartAutoDiscover();
        Task StopAutoDiscover();
        Task PublishSystemStatus(RedQueenSystemStatus status, string statusTopic);
        Task SubscribeSystemControlTopic(string controlTopic);
    }
    
    public class MqttService : IMqttService
    {
        private static Task OnPublisherConnected(MqttClientConnectedEventArgs evt)
        {
            System.Diagnostics.Debug.WriteLine("Publisher connected.");
            return Task.CompletedTask;
        }

        private static Task OnPublisherDisconnected(MqttClientDisconnectedEventArgs evt)
        {
            System.Diagnostics.Debug.WriteLine("Publisher disconnected.");
            return Task.CompletedTask;
        }

        private static Task OnSubscriberConnected(MqttClientConnectedEventArgs evt)
        {
            System.Diagnostics.Debug.WriteLine("Subscriber connected.");
            return Task.CompletedTask;
        }

        private static Task OnSubscriberDisconnected(MqttClientDisconnectedEventArgs evt)
        {
            System.Diagnostics.Debug.WriteLine("Subscriber disconnected.");
            return Task.CompletedTask;
        }

        private IManagedMqttClient _clientPublisher;
        private IManagedMqttClient _clientSubscriber;
        private readonly MqttBroker _broker;

        public event MqttMessageReceivedEventHandler MessageReceived;
        
        public bool IsDisposed { get; private set; }
        
        public string Host { get; private set; }
        
        public bool AutoDiscoverEnabled { get; private set; }

        public MqttService(MqttBroker broker)
        {
            _broker = broker;
            Host = broker.Host;
        }

        private void OnMessageReceived(MqttMessageReceivedEvent evt)
        {
            MessageReceived?.Invoke(this, evt);
        }

        private Task HandleReceivedReceivedMessage(MqttApplicationMessageReceivedEventArgs evt)
        {
            return Task.Run(() => OnMessageReceived(new MqttMessageReceivedEvent(evt, Host)));
        }

        private Task OnSubscriberMessageReceived(MqttApplicationMessageReceivedEventArgs evt)
        {
            return Task.Run(() => OnMessageReceived(new MqttMessageReceivedEvent(evt, Host)));
        }

        public ManagedMqttClientOptions GetOptions()
        {
            var keepAlive = _broker.KeepAliveSeconds ?? 5;

            var options = new ManagedMqttClientOptions
            {
                ClientOptions = new MqttClientOptions
                {
                    ClientId = "REDQUEEN",
                    ProtocolVersion = MqttProtocolVersion.V311,
                    ChannelOptions = new MqttClientTcpOptions
                    {
                        Server = _broker.Host,
                        Port = _broker.Port,
                        TlsOptions = new MqttClientTlsOptions
                        {
                            UseTls = _broker.UseTls,
                            IgnoreCertificateChainErrors = true,
                            IgnoreCertificateRevocationErrors = true,
                            AllowUntrustedCertificates = true
                        }
                    },
                    Credentials = new MqttClientCredentials(
                        _broker.Username,
                        Encoding.UTF8.GetBytes(_broker.Password)
                    ),
                    CleanSession = true,
                    KeepAlivePeriod = TimeSpan.FromSeconds(keepAlive)
                }
            };

            return options;
        }
        
        public async Task StartPublisher()
        {
            if (_clientPublisher != null)
            {
                return;
            }
            
            if (_broker == null)
            {
                throw new InvalidOperationException("MQTT Broker not defined.");
            }

            var options = GetOptions();
            if (options.ClientOptions.ChannelOptions == null)
            {
                throw new InvalidOperationException("Bad MQTT channel options");
            }
            
            var mqttFactory = new MqttFactory();
            
            _clientPublisher = mqttFactory.CreateManagedMqttClient();
            _clientPublisher.ApplicationMessageReceivedAsync += HandleReceivedReceivedMessage;
            _clientPublisher.ConnectedAsync += OnPublisherConnected;
            _clientPublisher.DisconnectedAsync += OnPublisherDisconnected;
            await _clientPublisher.StartAsync(options);
        }

        public async Task StopPublisher()
        {
            if (_clientPublisher == null)
            {
                return;
            }

            await _clientPublisher.StopAsync();
            _clientPublisher = null;
        }

        public async Task StartSubscriber()
        {
            if (_clientSubscriber != null)
            {
                return;
            }

            if (_broker == null)
            {
                throw new InvalidOperationException("MQTT Broker not defined.");
            }
            
            var options = GetOptions();
            if (options.ClientOptions.ChannelOptions == null)
            {
                throw new InvalidOperationException("Bad MQTT channel options");
            }
            
            var mqttFactory = new MqttFactory();
            _clientSubscriber = mqttFactory.CreateManagedMqttClient();
            _clientSubscriber.ConnectedAsync += OnSubscriberConnected;
            _clientSubscriber.DisconnectedAsync += OnSubscriberDisconnected;
            _clientSubscriber.ApplicationMessageReceivedAsync += OnSubscriberMessageReceived;
            await _clientSubscriber.StartAsync(options);
        }

        public async Task StopSubscriber()
        {
            if (_clientSubscriber == null)
            {
                return;
            }
            
            await _clientSubscriber.StopAsync();
            _clientSubscriber = null;
        }

        public async Task SubscribeTopic(MqttTopic topic)
        {
            if (topic.IsActive)
            {
                System.Diagnostics.Debug.WriteLine($"Subscribing topic: {topic.Name}");
                await _clientSubscriber.SubscribeAsync(topic.Name);
            }
        }

        public async Task SubscribeAllTopics()
        {
            foreach (var topic in _broker.Topics)
            {
                await SubscribeTopic(topic);
            }
        }

        public async Task StartAutoDiscover()
        {
            if (_broker == null)
            {
                throw new InvalidOperationException("MQTT Broker not defined.");
            }

            if (string.IsNullOrWhiteSpace(_broker.DiscoveryTopic))
            {
                System.Diagnostics.Debug.WriteLine("No Auto-Discover topic defined. Skipping auto-discover.");
                return;
            }

            if (AutoDiscoverEnabled)
            {
                return;
            }
            
            System.Diagnostics.Debug.WriteLine($"Discovery topic: {_broker.DiscoveryTopic}");
            
            var topic = new MqttTopic
            {
                Name = _broker.DiscoveryTopic.Trim(),
                BrokerId = _broker.Id,
                Broker = _broker,
                CreatedDate = DateTime.Now,
                IsActive = true
            };

            await SubscribeTopic(topic);
            AutoDiscoverEnabled = true;
        }

        public async Task StopAutoDiscover()
        {
            if (!AutoDiscoverEnabled || _broker == null || string.IsNullOrWhiteSpace(_broker.DiscoveryTopic))
            {
                return;
            }

            await _clientSubscriber.UnsubscribeAsync(_broker.DiscoveryTopic.Trim());
            AutoDiscoverEnabled = false;
        }

        public async Task PublishSystemStatus(RedQueenSystemStatus status, string statusTopic)
        {
            var payloadStr = JsonConvert.SerializeObject(status, new JsonSerializerSettings
            {
                ContractResolver = new DefaultContractResolver
                {
                    NamingStrategy = new CamelCaseNamingStrategy()
                },
                Formatting = Formatting.None
            });

            var msg = new MqttApplicationMessage
            {
                Topic = statusTopic,
                Payload = Encoding.UTF8.GetBytes(payloadStr),
                QualityOfServiceLevel = MqttQualityOfServiceLevel.AtMostOnce,
                Retain = true
            };

            await _clientPublisher.EnqueueAsync(msg);
        }

        public async Task SubscribeSystemControlTopic(string controlTopic)
        {
            var topic = new MqttTopic
            {
                Name = controlTopic,
                IsActive = true,
                Broker = _broker,
                BrokerId = _broker.Id,
                CreatedDate = DateTime.Now
            };

            await SubscribeTopic(topic);
        }
        
        public void Dispose()
        {
            if (IsDisposed)
            {
                return;
            }

            StopAutoDiscover().Wait();
            StopPublisher().Wait();
            StopSubscriber().Wait();
            IsDisposed = true;
            GC.SuppressFinalize(this);
        }
    }
}

File: redqueen-backend/redqueen/src/mqtt/mqtt.service.ts
import {
  MqttBroker,
  MqttTopic,
  RedqueenSystemStatusDto,
  Utils,
} from '@redqueen-backend/redqueen-data';
import mqtt, {
  IConnackPacket,
  IDisconnectPacket,
  IPublishPacket,
  MqttClient,
} from 'mqtt';
import { Logger } from 'winston';
import {
  MqttMessageReceivedArgs,
  MqttServiceEventCallbacks,
} from '../events/mqtt-service-event-callbacks';
import { TypedEventEmitter } from '../events/event-emitter';

export class MqttService extends TypedEventEmitter<MqttServiceEventCallbacks> {
  private readonly _broker: MqttBroker;
  private readonly _host: string;
  private readonly _logger: Logger;
  private _autoDiscoverEnabled: boolean;
  private _isDisposed: boolean;
  private _clientPublisher?: MqttClient;
  private _clientSubscriber?: MqttClient;

  constructor(logger: Logger, broker: MqttBroker) {
    super();
    this._logger = logger;
    this._broker = broker;
    this._host = `mqtt://${this._broker.host}`;
    this._autoDiscoverEnabled = false;
    this._isDisposed = false;
  }

  public get host(): string {
    return this._host;
  }

  public get isDisposed(): boolean {
    return this._isDisposed;
  }

  public get autoDiscoverEnabled(): boolean {
    return this._autoDiscoverEnabled;
  }

  public getOptions(): mqtt.IClientOptions {
    const opts: mqtt.IClientOptions = {
      clientId: 'REDQUEEN',
      protocolVersion: 4,
      port: this._broker.port,
      clean: true,
      keepalive: this._broker.keepAliveSeconds || 5,
      protocolId: 'MQTT',
      username: this._broker.username,
      password: Buffer.from(this._broker.password),
    };
    return opts;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private onPublisherConnected(packet: IConnackPacket): void {
    this._logger.info('Publisher connected');
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private onPublisherDisconnected(packet: IDisconnectPacket): void {
    this._logger.info('Publisher disconnected');
  }

  private onPublisherMessageReceived(
    topic: string,
    payload: Buffer,
    packet: IPublishPacket,
  ): void {
    const args: MqttMessageReceivedArgs = {
      topic: topic,
      payload: payload,
      packet: packet,
      host: this._host,
    };
    this.emit('messageReceived', this, args);
  }

  public async startPublisher(): Promise<void> {
    if (this._clientPublisher || this._isDisposed) {
      return;
    }

    const opts = this.getOptions();
    this._clientPublisher = await mqtt.connectAsync(this.host, opts);
    this._clientPublisher.on('connect', (packet: IConnackPacket) =>
      this.onPublisherConnected(packet),
    );
    this._clientPublisher.on('disconnect', (packet: IDisconnectPacket) =>
      this.onPublisherDisconnected(packet),
    );
    this._clientPublisher.on(
      'message',
      (topic: string, payload: Buffer, packet: IPublishPacket) =>
        this.onPublisherMessageReceived(topic, payload, packet),
    );
  }

  public async stopPublisher(): Promise<void> {
    if (!this._clientPublisher || this._isDisposed) {
      return;
    }

    this._clientPublisher.removeAllListeners();
    await this._clientPublisher.endAsync();
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private onSubscriberConnected(packet: IConnackPacket): void {
    this._logger.info('Subscriber connected');
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private onSubscriberDisconnected(packet: IDisconnectPacket): void {
    this._logger.info('Subscriber disconnected');
  }

  private onSubscriberMessageReceived(
    topic: string,
    payload: Buffer,
    packet: IPublishPacket,
  ): void {
    const args: MqttMessageReceivedArgs = {
      topic: topic,
      payload: payload,
      packet: packet,
      host: this._broker.host,
    };
    this.emit('messageReceived', this, args);
  }

  public async startSubscriber(): Promise<void> {
    if (this._clientSubscriber || this._isDisposed) {
      return;
    }

    const opts = this.getOptions();
    this._clientSubscriber = await mqtt.connectAsync(this.host, opts);
    this._clientSubscriber.on('connect', (packet: IConnackPacket) =>
      this.onSubscriberConnected(packet),
    );
    this._clientSubscriber.on('disconnect', (packet: IDisconnectPacket) =>
      this.onSubscriberDisconnected(packet),
    );
    this._clientSubscriber.on(
      'message',
      (topic: string, payload: Buffer, packet: IPublishPacket) =>
        this.onSubscriberMessageReceived(topic, payload, packet),
    );
  }

  public async stopSubscriber(): Promise<void> {
    if (!this._clientSubscriber || this._isDisposed) {
      return;
    }

    this._clientSubscriber.removeAllListeners();
    await this._clientSubscriber.endAsync();
  }

  public async subscribeTopic(topic: MqttTopic): Promise<void> {
    if (topic.isActive) {
      this._logger.info(`Subscribing topic: ${topic.name}`);
      await this._clientSubscriber.subscribeAsync(topic.name, {
        qos: 2,
      });
    }
  }

  public async subscribeAllTopics(): Promise<void> {
    console.log('topics:', this._broker.topics);
    for (const topic of this._broker.topics) {
      await this.subscribeTopic(topic);
    }
  }

  public async startAutoDiscover(): Promise<void> {
    if (this.autoDiscoverEnabled || this._isDisposed) {
      return;
    }

    if (Utils.isNullOrWhitespace(this._broker.discoveryTopic)) {
      this._logger.info(
        'No Auto-Discover topic defined. Skipping auto-discover.',
      );
      return;
    }

    this._logger.info(`Discovery topic: ${this._broker.discoveryTopic}`);
    const topic = new MqttTopic();
    topic.name = this._broker.discoveryTopic.trim();
    topic.brokerId = this._broker.id;
    topic.broker = this._broker;
    topic.createdDate = new Date();
    topic.isActive = true;

    await this.subscribeTopic(topic);
    this._autoDiscoverEnabled = true;
  }

  public async stopAutoDiscover(): Promise<void> {
    if (
      !this._autoDiscoverEnabled ||
      Utils.isNullOrWhitespace(this._broker.discoveryTopic) ||
      this._isDisposed
    ) {
      return;
    }

    await this._clientSubscriber.unsubscribeAsync(
      this._broker.discoveryTopic.trim(),
    );
    this._autoDiscoverEnabled = false;
  }

  public async publishPayloadToTopic(
    payload: any,
    topic: string,
  ): Promise<void> {
    const payloadStr = JSON.stringify(payload);
    this._logger.info(`Publishing payload: ${payload} to topic: ${topic} ...`);
    await this._clientPublisher.publishAsync(topic, payloadStr, {
      qos: 2,
      retain: true,
    });
  }

  public async publishSystemStatus(
    status: RedqueenSystemStatusDto,
    statusTopic: string,
  ): Promise<void> {
    return await this.publishPayloadToTopic(status, statusTopic);
  }

  public async subscribeSystemControlTopic(
    controlTopic: string,
  ): Promise<void> {
    const topic = new MqttTopic();
    topic.name = controlTopic;
    topic.broker = this._broker;
    topic.brokerId = this._broker.id;
    topic.createdDate = new Date();
    await this.subscribeTopic(topic);
  }

  public async dispose(): Promise<void> {
    if (this.isDisposed) {
      return;
    }

    await this.stopAutoDiscover();
    await this.stopPublisher();
    await this.stopSubscriber();
    this._isDisposed = true;
  }
}
